CMakeLists.txt

```
@@ -0,0 +1,297 @@
+  enabled in Monitor (that is Monitor::is_active(\"DEBUG\") must be true for any output)" OFF)
+option(debug_field "" OFF)
+option(debug_field_face "" OFF)
+if (debug)
+  add_compile_definitions(CELLO_DEBUG)
+endif()
+if (debug_field)
+  add_compile_definitions(DEBUG_FIELD)
+endif()
+if (debug_field_face)
+  add_compile_definitions(DEBUG_FIELD_FACE)
+endif()
+
+
+option(check "Do extra run-time checking.  Useful for debugging, but can potentially slow calculations down" OFF)
+if (check)
+  add_compile_definitions(CELLO_CHECK)
+endif()
+
+
+option(debug_verbose "Print periodically all field values.  See src/Field/field_FieldBlock.cpp" OFF)
+if (debug_verbose)
+  add_compile_definitions(CELLO_DEBUG_VERBOSE)
+endif()
+
+option(memory "Track dynamic memory statistics.  Can be useful, but can cause problems on some \
+  systems that also override new [] () / delete [] ()" OFF)
+if (memory)
+  add_compile_definitions(CONFIG_USE_MEMORY)
+endif()
+
+option(balance "Enable charm++ dynamic load balancing" ON)
+set(balancer_included "CommonLBs" CACHE STRING "Charm++ load balancer to include")
+set(balancer_default "TreeLB" CACHE STRING "Charm++ load balancer to use by default")
+#TODO we should figure our reasonale defaults and/or provide instructions in the docs
+if (balance)
+  foreach(BALANCER IN LISTS balancer_included)
+    string(APPEND Cello_TARGET_LINK_OPTIONS " -module ${BALANCER}")
+  endforeach()
+  foreach(BALANCER IN LISTS balancer_default)
+    string(APPEND Cello_TARGET_LINK_OPTIONS " -balancer ${BALANCER}")
+  endforeach()
+endif()
+
+
+option(use_gprof "Compile with -pg to use gprof for performance profiling" OFF)
+if (use_gprof)
+  SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -pg")
+  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pg")
+  SET(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -pg")
+endif()
+
+
+option(use_valgrind "Run the test programs using valgrind to check for memory leaks" OFF)
+if ()
+  add_compile_definitions()
+endif()
+
+
+option(use_performance "Use Cello Performance class for collecting performance \
+  data (currently requires global reductions, and may not be fully \
+  functional) (basic time data on root processor is still output)" ON)
+if (use_performance)
+  add_compile_definitions(CONFIG_USE_PERFORMANCE)
+endif()
+
+
+option(use_projections "Compile the CHARM++ version for use with the Projections performance tool." OFF)
+if (use_projections)
+  add_compile_definitions(CONFIG_USE_PROJECTIONS)
+  string(APPEND Cello_TARGET_LINK_OPTIONS " -tracemode projections")
+endif()
+
+
+set(ip_charm "4" CACHE STRING "Number of processors to run parallel unit tests (legacy name)")
+set(PARALLEL_LAUNCHER ${CHARM_RUN} CACHE STRING "Launcher to use for parallel tests")
+set(PARALLEL_LAUNCHER_NPROC_ARG "+p" CACHE STRING "Argument to set number of processing elements for parallel launcher")
+set(PARALLEL_LAUNCHER_NPROC ${ip_charm} CACHE STRING "Number of processors to run parallel unit tests")
+
```

The following recommendation enforces an assumption about the value of ``PARALLEL_LAUNCHER_NPROC_ARG`` that is required for all parallel tests to work (this tripped me up a little)

```suggestion
if ("${PARALLEL_LAUNCHER_NPROC_ARG}" MATCHES "[ \t\r\n]")
  message(FATAL_ERROR
    "Whitespace should NOT be present in PARALLEL_LAUNCHER_NPROC_ARG.\n"
    "If you want to use PARALLEL_LAUNCHER_NPROC_ARG to pass multiple "
    "arguments to the launcher, use a semicolon to delimit each argument.\n"
    "For example, instead of setting PARALLEL_LAUNCHER_NPROC_ARG to "
    "\"++local +p\", set it equal to \"++local;+p\".")
endif()

```

---

Good point. I hadn't tested multiple arguments yet.